core/src/main/java/org/infinispan/cache/impl/CacheBackedEntrySet.java

```diff
@@ -0,0 +1,65 @@
+package org.infinispan.cache.impl;
+
+import java.util.Map;
+import java.util.function.Function;
+
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.container.entries.ForwardingCacheEntry;
+import org.infinispan.context.InvocationContext;
+import org.infinispan.stream.StreamMarshalling;
+
+/**
+ * Entry set backed by a cache.
+ *
+ * <p>Implements {@code CacheSet<CacheEntry<K, V>>} but it is also (mis-)used as a {@code CacheSet<Map.Entry<K, V>>}.
+ * This works because {@code add()} and {@code addAll()} are not implemented.</p>
+ *
+ * @since 13.0
+ */
+public class CacheBackedEntrySet<K, V> extends AbstractCacheBackedSet<K, V, CacheEntry<K, V>> {
+   public CacheBackedEntrySet(CacheImpl<K, V> cache, Object lockOwner, long explicitFlags) {
+      super(cache, lockOwner, explicitFlags);
+   }
+
+   @Override
+   public boolean contains(Object o) {
+      if (!(o instanceof Map.Entry))
+         return false;
+
+      Map.Entry<?, ?> entry = (Map.Entry<?, ?>) o;
+
+      InvocationContext ctx = cache.invocationContextFactory.createInvocationContext(false, 1);
+      V cacheValue = cache.get(entry.getKey(), explicitFlags, ctx);
+
+      return cacheValue != null && cacheValue.equals(entry.getValue());
+   }
+
+   @Override
+   protected Function<Map.Entry<K, V>, ?> entryToKeyFunction() {
+      return StreamMarshalling.entryToKeyFunction();
+   }
+
+   @Override
+   protected Object extractKey(Object e) {
+      if (!(e instanceof Map.Entry))
+         return null;
+
+      return ((Map.Entry<?, ?>) e).getKey();
+   }
+
+   @Override
+   protected CacheEntry<K, V> wrapElement(CacheEntry<K, V> e) {
+      return new ForwardingCacheEntry<K, V>() {
```

Comment:
REMOTE_ITERATION wouldn't be set on the origin node, so yeah it was the same before.

But I see what you mean, `forEach()` and `removeIf` could bypass the operator and avoid wrapping each entry because they don't need `setValue()`.

Reply:
Done